### PR TITLE
Add check for duplicate JSON keys - v1

### PR DESCRIPTION
**Pull request by jasonish**
Add a new script to test for duplicate JSON keys.

TODO: Hook info normal S-V runs somehow.

### CODE REVIEW

**Comment by victorjulien** [comment]
I'm fine with this as a first step, but really feel it should be managed by `run.py` so we don't need to do the additional step of running this check. Probably needs master-6.0.x to be clean first.

**Comment by jasonish** [comment]
> I'm fine with this as a first step, but really feel it should be managed by `run.py` so we don't need to do the additional step of running this check. Probably needs master-6.0.x to be clean first.

Yeah, makes sense.  Probably as part of eve-validation.  Will come back to that soon, integrating properly is a bigger job than creating this script! :)

**Comment by catenacyber** [comment]
It is no longer requires suricata fix, but requires backports, right ?

**Comment by jasonish** [comment]
> It is no longer requires suricata fix, but requires backports, right ?

Correct.  Lets just put this on hold til after these releases are out.

**Comment by catenacyber** [comment]
> Correct. Lets just put this on hold til after these releases are out.

Ok.
By the way, why is CI green if a backport is still missing ?

**Comment by jasonish** [comment]
> > Correct. Lets just put this on hold til after these releases are out.
> 
> Ok. By the way, why is CI green if a backport is still missing ?

This script isn't hooked into this CI yet, for now its just a utility script.  My original PR message is in error.

**Comment by catenacyber** [comment]
Should not jsonschema tell us about duplicate keys ?

**Comment by catenacyber** [comment]
> No, jsonschema runs on the data after loaded by Python. The Python loader throws out duplicate keys.

jsonschema runs on the file directly cf https://github.com/OISF/suricata-verify/blob/master/eve-validator/src/main.rs#L52

Now that it is the fast rust version instead of the slow python one

**Comment by jasonish** [comment]
> Can we disallow dups for eve anyway? It's almost certainly unintented?

Verification is probably the best way (ie: this PR),  It would be hard to do with JsonBuilder given we're not inserting to a map that might know what keys are already there, so its state tracking would have to become much more advanced.

**Comment by jasonish** [comment]
aaaasdfffttrreewwqqzzxxcVVVVR

> > Can we disallow dups for eve anyway? It's almost certainly unintented?
> 
> Verification is probably the best way (ie: this PR), It would be hard to do with JsonBuilder given we're not inserting to a map that might know what keys are already there, so its state tracking would have to become much more advanced.

Alternative answer.. Of course, its just software, but not a quick or easy change to JsonBuilder.

**Comment by victorjulien** [comment]
Right, I was thinking about verification stage indeed.

**Comment by catenacyber** [comment]
And so, verification with serde would not tell us about duplicate keys ?

**Comment by jasonish** [comment]
> And so, verification with serde would not tell us about duplicate keys ?

Not easily.  Using serde to decode to a generate `serde_json::Value` will not show these errors.

If you create structs for all possible data types and derive the serde implementation, it will error out on duplicate keys.  Maybe we could generate a struct from the schema and deserialize every line to it?  I'm not sure how robust the existing, if any jsonschema -> rust code gen tools there are though.

**Comment by catenacyber** [comment]
Oh, so python errors on duplicate keys but not rust `serde_json` 😢 

So, this is a good idea

**Comment by jasonish** [comment]
> Oh, so python errors on duplicate keys but not rust `serde_json` cry

Python doesn't either. This script hooks into the deserializer to raise the error.

File: duplicate-key-check.py [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+            print("ERROR: Unknown: {}: {}".format(filename, err))
+            failures += 1
+
+    return failures
+
+def hook(pairs):
+    d = {}
+    for k, v in pairs:
+        if k in d:
+            equal = d[k] == v
```
Comment (catenacyber) [inline duplicate-key-check.py:36]: This seems redundant with the warning message below

File: duplicate-key-check.py [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+        for line in json_in:
+            json.loads(line, object_pairs_hook=hook)
+
+def main():
+    failures = 0
+    for filename in sys.argv[1:]:
+        try:
+            validate_file(filename)
+        except DuplicateKeyError as err:
+            print("ERROR: Duplicate key: {}: {}".format(filename, err))
```
Comment (catenacyber) [inline duplicate-key-check.py:25]: Should we increase failures here as well ?

File: duplicate-key-check.py [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+def hook(pairs):
+    d = {}
+    for k, v in pairs:
+        if k in d:
+            equal = d[k] == v
+            raise DuplicateKeyError("key={}, equal={} current value={}, new value={}".format(k, equal, d[k], v))
+        d[k] = v
+    return d
+
+if __name__ == "__main__":
```
Comment (catenacyber) [inline duplicate-key-check.py:41]: Who calls this ?

File: duplicate-key-check.py (hunk 1 above)
Comment (jasonish) [reply duplicate-key-check.py:36]: Ah, yeah.. Debugging.

File: duplicate-key-check.py (hunk 3 above)
Comment (jasonish) [reply duplicate-key-check.py:41]: We'd have to hook it in, this is just a script for detecting duplicates.

**Comment by catenacyber** [comment]
Tests pass because the new script is not hooked in yet

**Comment by victorjulien** [comment]
> Tests pass because the new script is not hooked in yet

Updated my script to exclude drafts.

**Comment by catenacyber** [comment]
@jasonish what is the status if this ?
This looks a good idea, what is missing before merging ?

**Comment by jasonish** [comment]
Replaced by https://github.com/OISF/suricata-verify/pull/1572.
